services/MerchStoreService.ts

```diff
@@ -291,6 +291,68 @@ export default class MerchStoreService {
       await user.reload();
       const merchItemOptionRepository = Repositories.merchStoreItemOption(txn);
       const itemOptions = await merchItemOptionRepository.batchFindByUuid(originalOrder.map((oi) => oi.option));
+
+      const totalCost = await this.verifyOrder(originalOrder, user);
+
+      const merchOrderRepository = Repositories.merchOrder(txn);
+
+      // if all checks pass, the order is placed
+      const createdOrder = await merchOrderRepository.createMerchOrder(OrderModel.create({
+        user,
+        totalCost,
+        items: flatten(originalOrder.map((optionAndQuantity) => {
+          const option = itemOptions.get(optionAndQuantity.option);
+          const quantityRequested = optionAndQuantity.quantity;
+          return Array(quantityRequested).fill(OrderItemModel.create({
+            option,
+            salePriceAtPurchase: option.getPrice(),
+            discountPercentageAtPurchase: option.discountPercentage,
+          }));
+        })),
+      }));
+
+      const activityRepository = Repositories.activity(txn);
+      await activityRepository.logActivity({
+        user,
+        type: ActivityType.ORDER_MERCHANDISE,
+        description: `Order ${createdOrder.uuid}`,
+      });
+
+      await Promise.all(originalOrder.map(async (optionAndQuantity) => {
+        const option = itemOptions.get(optionAndQuantity.option);
+        const updatedQuantity = option.quantity - optionAndQuantity.quantity;
+        return merchItemOptionRepository.upsertMerchItemOption(option, { quantity: updatedQuantity });
+      }));
+
+      const userRepository = Repositories.user(txn);
+      userRepository.upsertUser(user, { credits: user.credits - totalCost });
+
+      return [createdOrder, itemOptions];
+    });
+
+    const orderConfirmation = {
+      items: originalOrder.map((oi) => {
+        const option = merchItemOptions.get(oi.option);
+        const { item } = option;
+        return {
+          ...item,
+          quantityRequested: oi.quantity,
+          salePrice: option.getPrice(),
+          total: oi.quantity * option.getPrice(),
+        };
+      }),
+      totalCost: order.totalCost,
+    };
+    this.emailService.sendOrderConfirmation(user.email, user.firstName, orderConfirmation);
+
+    return order.getPublicOrder();
+  }
+
+  public async verifyOrder(originalOrder: MerchItemOptionAndQuantity[], user: UserModel): Promise<number> {
```

Comment:
`placeOrder` creates a transaction and calls `verifyOrder` from within that transaction. `verifyOrder` also creates a transaction though. Not sure if `verifyOrder` uses the `placeOrder` transaction or if we end up in a deadlock or what happens (tests pass though so I assume it's the former), but we should have two `verifyOrder` methods I think: one that's called from the controller and creates a transaction, and another that accepts a transaction as a param and is called by `placeOrder` and the other `verifyOrder`.

Reply:
Are there any good resources on how to pass transactions as parameters? Haven't been able to find something that applies to this situation

Reply:
What we're trying to do here is restructure the code so that we don't try opening a transaction within a transaction. What's going on right now is
```typescript
// the controller calls this
async placeOrder() {
  this.transactions.readWrite(txn -> {
    verifyOrder();
  });
}

// the controller and placeOrder call this
async verifyOrder() {
  this.transactions.readWrite(txn -> {
    // do something
  });
}
```
`placeOrder` creates a transaction and calls `verifyOrder` from within that transaction. `verifyOrder` also creates a transaction though, so we try opening a transaction within a transaction. Best case, the code's a little confusing to read but the library we use to call our database handles it intelligently and doesn't try opening up another transaction (which seems to be the case); worst case, something gets stuck and prevents the backend from crashing and from processing other requests.

Instead, we should have two `verifyOrder` methods:
```typescript
// the controller calls this
async placeOrder() {
  this.transactions.readWrite(txn -> {
    verifyOrderInternal(txn);
  });
}

// the controller calls this
async verifyOrder() {
  this.transactions.readWrite(txn -> {
    verifyOrderInternal(txn);
  });
}

// placeOrder and verifyOrder call this
async verifyOrderInternal(txn: TransactionalEntityManager) {
 // do something
}
```
Here's an example from our repo that passes the transaction as a method parameter: [`TransactionsManager`](https://github.com/acmucsd/membership-portal/blob/master/repositories/index.ts#L64). Going off that, I think `TransactionalEntityManager` is the correct type, but you can look at the type definitions in your IDE to verify.